<redacted> <redacted>
https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>
<redacted>
1315700

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely or locally)
flower_data = pd.read_csv("iris.csv")
# Output the first 15 rows of the data
print(flower_data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(flower_data.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data is using different features such as different measurements of part of the flower to label what species of Iris these plants are. The features are the sepal_length, sepal_width, petal_length, petal_width and the labels are the values in the species column. The mapping of the the labels to the actual classes would just be the species of the Iris where each class is just the label added to Iris, so one class would just be Iris Setosa as shown in the first 15 of the dataset.

# Part 2: Split the dataset into train and test

In [3]:
# Take the dataset and split it into our features (X) and label (y)
labels = flower_data.iloc[:,-1]

features = flower_data.iloc[:,:-1]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train , X_test , y_train, y_test = train_test_split(features, labels, test_size=0.4, train_size=0.6, random_state=16)


# Part 3: Logistic Regression

In [4]:
# i. Use sklearn to train a LogisticRegression model on the training set
LRmodel = LogisticRegression()
LRmodel.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
LRpredict = LRmodel.predict_proba(X_test)
print(LRpredict)
# iii. Report on the score for Logistic regression model, what does the score measure?
LRscore = LRmodel.score(X_test, y_test)
print(LRscore)
# iv. Extract the coefficents and intercepts for the boundary line(s)
boundaryC = LRmodel.coef_
print(boundaryC)
boundaryI = LRmodel.intercept_
print(boundaryI)


[[5.31272120e-03 5.81857324e-01 4.12829954e-01]
 [9.83591664e-01 1.64080784e-02 2.57593192e-07]
 [3.81823318e-03 7.28205095e-01 2.67976672e-01]
 [9.47070980e-04 3.74359613e-01 6.24693316e-01]
 [4.57366502e-03 6.65058567e-01 3.30367768e-01]
 [9.62511537e-01 3.74874890e-02 9.74142980e-07]
 [8.42084351e-03 8.18194061e-01 1.73385095e-01]
 [9.58949835e-01 4.10488230e-02 1.34206359e-06]
 [9.74457537e-01 2.55419216e-02 5.40948993e-07]
 [1.93017819e-02 8.55335413e-01 1.25362805e-01]
 [6.47612610e-02 9.18004463e-01 1.72342764e-02]
 [1.07974827e-04 5.05072648e-02 9.49384760e-01]
 [1.40910374e-04 1.93871432e-01 8.05987658e-01]
 [9.58096360e-01 4.19028175e-02 8.22902710e-07]
 [1.45135089e-04 1.21685783e-01 8.78169082e-01]
 [2.05331464e-02 8.90168274e-01 8.92985795e-02]
 [9.67341012e-01 3.26582323e-02 7.56012444e-07]
 [9.58153787e-01 4.18442613e-02 1.95162082e-06]
 [7.16898147e-03 5.11068656e-01 4.81762362e-01]
 [4.73159615e-03 8.28569557e-01 1.66698847e-01]
 [9.65083261e-01 3.49159541e-02 7.853721

The score represents that there is about a 95% mean accuracy in the models predictions

# Part 4: Support Vector Machine

In [5]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVMmodel = svm.SVC(probability=True)
SVMmodel.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
SVCpredict = SVMmodel.predict_proba(X_test)
print(SVCpredict)
# iii. Report on the score for the SVM, what does the score measure?
SVMscore = SVMmodel.score(X_test, y_test)
print(SVMscore)

[[0.01405367 0.74322966 0.24271667]
 [0.96998677 0.01738676 0.01262647]
 [0.01526879 0.80987435 0.17485686]
 [0.01412105 0.31629684 0.66958211]
 [0.01107615 0.78655633 0.20236753]
 [0.95270737 0.03335468 0.01393795]
 [0.01232849 0.91951662 0.06815489]
 [0.94955704 0.03602301 0.01441996]
 [0.96307063 0.02421747 0.01271189]
 [0.01068516 0.96009102 0.02922382]
 [0.0291591  0.96145632 0.00938457]
 [0.01205905 0.02808307 0.95985788]
 [0.01364603 0.07319079 0.91316318]
 [0.95448095 0.03189015 0.01362889]
 [0.01189659 0.05223136 0.93587205]
 [0.01062592 0.96710436 0.02226971]
 [0.95958775 0.0272946  0.01311765]
 [0.94540014 0.03943796 0.0151619 ]
 [0.01734262 0.72855994 0.25409744]
 [0.00886317 0.89456213 0.0965747 ]
 [0.95831557 0.02835895 0.01332549]
 [0.96779765 0.01987972 0.01232262]
 [0.01514396 0.29404983 0.69080621]
 [0.01354877 0.00542707 0.98102416]
 [0.01256099 0.8827081  0.10473092]
 [0.0125675  0.96845933 0.01897318]
 [0.01301166 0.6743368  0.31265154]
 [0.01100051 0.01161759 0.97

The score represents that there is about a 95% mean accuracy in the models predictions

# Part 5: Neural Network

In [6]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NNmodel = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=6)
NNmodel.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
NNpredict = NNmodel.predict_proba(X_test)
print(NNpredict)
# iii. Report on the score for the Neural Network, what does the score measure?
NNScore = NNmodel.score(X_test, y_test)
print(NNScore)
# iv: Experiment with different options for the neural network, report on your best configuration

[[0.00000000e+000 1.00000000e+000 2.25630547e-264]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.00000000e+000 7.02745157e-139]
 [0.00000000e+000 1.00000000e+000 2.38541142e-172]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.00000000e+000 0.00000000e+000]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.00000000e+000 1.32086274e-223]
 [0.00000000e+000 1.00000000e+000 1.91817706e-272]
 [0.00000000e+000 1.38220527e-293 1.00000000e+000]
 [0.00000000e+000 1.32970998e-295 1.00000000e+000]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.51937553e-176 1.00000000e+000]
 [0.00000000e+000 1.00000000e+000 1.51579791e-175]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [1.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 3.94525932e-310 1.00000000e+000]
 [0.00000000e+000 1.00000000e+0

The score represents that there is about a 98.3% mean accuracy in the models predictions

After trying out multiple configurations, it seems like solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=6 is what was the most accurate for me. the random state went up as it approached 6-8 and after 8 went back down, the hidden_layer worked best at 5,2 and the alpha worked best for anything what was 1e-x where x was greater than 5

# Part 6: K-Nearest Neighbors

In [7]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
KNNmodel = KNeighborsClassifier(n_neighbors=3)
KNNmodel.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
KNNpredict = KNNmodel.predict_proba(X_test)
print(KNNpredict)
# iii. Report on the score for kNN, what does the score measure?
KNNscore = KNNmodel.score(X_test, y_test)
print(KNNscore)

[[0.         0.66666667 0.33333333]
 [1.         0.         0.        ]
 [0.         1.         0.        ]
 [0.         0.66666667 0.33333333]
 [0.         0.33333333 0.66666667]
 [1.         0.         0.        ]
 [0.         1.         0.        ]
 [1.         0.         0.        ]
 [1.         0.         0.        ]
 [0.         1.         0.        ]
 [0.         1.         0.        ]
 [0.         0.         1.        ]
 [0.         0.         1.        ]
 [1.         0.         0.        ]
 [0.         0.         1.        ]
 [0.         1.         0.        ]
 [1.         0.         0.        ]
 [1.         0.         0.        ]
 [0.         0.33333333 0.66666667]
 [0.         1.         0.        ]
 [1.         0.         0.        ]
 [1.         0.         0.        ]
 [0.         0.33333333 0.66666667]
 [0.         0.         1.        ]
 [0.         1.         0.        ]
 [0.         1.         0.        ]
 [0.         1.         0.        ]
 [0.         0.         1.  

The score represents that there is about a 96.7% mean accuracy in the models predictions

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

From the results I achieved, it seems as though all of the models are able to test at a relatively high accuracy on this specific dataset. The best performing for me was the Neural Network after using an optomized configuration. I think this was the best as it was able to be more accurately fine tuned for the data it was given. Something that surprised me was that when I had used the original random seed of 42, I was getting only 1.0 values, but by changing it to another value like 16 I was able to get more variable and descriptive data to measure. I even tested other seeds and my results were more descriptive closer to the 16 value.